File: Algorithms/src/1594. Maximum Non Negative Product in a Matrix/Code/solution.java
class Solution {
    public int maxProductPath(int[][] grid) {
        int m = grid.length, n = grid[0].length;
        long MOD = (long)1e9 + 7;

        long[][] maxDp = new long[m][n];
        long[][] minDp = new long[m][n];

        maxDp[0][0] = minDp[0][0] = grid[0][0];

        for (int i = 1; i < m; i++) {
            maxDp[i][0] = minDp[i][0] = maxDp[i-1][0] * grid[i][0];
        }

        for (int j = 1; j < n; j++) {
            maxDp[0][j] = minDp[0][j] = maxDp[0][j-1] * grid[0][j];
        }

        for (int i = 1; i < m; i++) {
            for (int j = 1; j < n; j++) {
                long val = grid[i][j];

                long a = maxDp[i-1][j] * val;
                long b = minDp[i-1][j] * val;
                long c = maxDp[i][j-1] * val;
                long d = minDp[i][j-1] * val;

                maxDp[i][j] = Math.max(Math.max(a, b), Math.max(c, d));
                minDp[i][j] = Math.min(Math.min(a, b), Math.min(c, d));
            }
        }

        long res = maxDp[m-1][n-1];
        return res < 0 ? -1 : (int)(res % MOD);
    }
}

File: Algorithms/src/1594. Maximum Non Negative Product in a Matrix/Code/solution.py
class Solution:
    def maxProductPath(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        MOD = 10**9 + 7

        maxDp = [[0]*n for _ in range(m)]
        minDp = [[0]*n for _ in range(m)]

        maxDp[0][0] = minDp[0][0] = grid[0][0]

        for i in range(1, m):
            maxDp[i][0] = minDp[i][0] = maxDp[i-1][0] * grid[i][0]

        for j in range(1, n):
            maxDp[0][j] = minDp[0][j] = maxDp[0][j-1] * grid[0][j]

        for i in range(1, m):
            for j in range(1, n):
                val = grid[i][j]

                candidates = [
                    maxDp[i-1][j] * val,
                    minDp[i-1][j] * val,
                    maxDp[i][j-1] * val,
                    minDp[i][j-1] * val
                ]

                maxDp[i][j] = max(candidates)
                minDp[i][j] = min(candidates)

        res = maxDp[m-1][n-1]
        return -1 if res < 0 else res % MOD